test/libgit2.jl

```diff
@@ -1549,19 +1578,76 @@ mktempdir() do dir
+                # User provides a valid username
+                challenges = [
+                    "Username for 'github.com':" => "$username\n",
+                ]
+                err, auth_attempts = challenge_prompt(ssh_u_cmd, challenges)
+                @test err == 0
+                @test auth_attempts == 1
+
+                # User sends EOF in username prompt which aborts the credential request
+                challenges = [
+                    "Username for 'github.com':" => "\x04",
+                ]
+                err, auth_attempts = challenge_prompt(ssh_u_cmd, challenges)
+                @test err == abort_prompt
+                @test auth_attempts == 1
+
+                # User provides an empty username
+                challenges = [
+                    "Username for 'github.com':" => "\n",
+                    "Username for 'github.com':" => "\x04",
+                ]
+                err, auth_attempts = challenge_prompt(ssh_u_cmd, challenges)
+                @test err == abort_prompt
+                @test auth_attempts == 5  # Should ideally be <= 2
+
+                # User repeatedly chooses an invalid username
+                challenges = [
+                    "Username for 'github.com':" => "foo\n",
+                    "Username for 'github.com' [foo]:" => "\n",
+                    "Username for 'github.com' [foo]:" => "\x04",  # Need to manually abort
+                ]
+                err, auth_attempts = challenge_prompt(ssh_u_cmd, challenges)
+                @test err == abort_prompt
+                @test auth_attempts == 6
+
+                # Credential callback is given an empty string in the `username_ptr`
+                # instead of the typical C_NULL.
+                ssh_user_empty_cmd = """
+                include("$LIBGIT2_HELPER_PATH")
+                valid_cred = LibGit2.SSHCredentials("$username", "", "$valid_key", "$valid_key.pub")
+                err, auth_attempts = credential_loop(valid_cred, "$url", "")
+                (err < 0 ? LibGit2.GitError(err) : err, auth_attempts)
+                """
+
+                challenges = [
+                    "Username for 'github.com':" => "$username\n",
+                ]
+                err, auth_attempts = challenge_prompt(ssh_user_empty_cmd, challenges)
+                @test err == 0
+                @test auth_attempts == 1
+            end
+
             # TODO: Tests are currently broken. Credential callback prompts for:
             # "Passphrase for :"
-            #=
             # Explicitly setting these env variables to be empty means the user will be
             # given a prompt with no defaults set.
-            withenv("SSH_KEY_PATH" => "", "SSH_PUB_KEY_PATH" => "") do
+            withenv("SSH_KEY_PATH" => nothing,
+                    "SSH_PUB_KEY_PATH" => nothing,
+                    "SSH_KEY_PASS" => nothing,
+                    (Sys.iswindows() ? "USERPROFILE" : "HOME") => tempdir()) do
```

Comment:
is this relying on tempdir happening to not have any key files in it? probably unlikely, but never know